SimPEG/directives/directives.py

```diff
@@ -386,13 +389,13 @@ def seed(self):
 
         Returns
         -------
-        int
+        int, numpy.random.Generator or None
         """
         return self._seed
 
     @seed.setter
     def seed(self, value):
-        if value is not None:
+        if value is not None and not isinstance(value, np.random.Generator):
```

Comment:
Maybe we should instead do,
```Python
try:
    rng = np.random.default_generator(value)
except TypeError as err:
    raise TypeError( f'Unable to initialize the random number generator with a {type(value).__name__}`) from err

self._rng = rng
self._seed = rng.bit_generator.seed_seq
```

Which will allow someone to both see the actual used seed (If `None` was passed), and re-use it for a future run.

Reply:
Not sure what to do if someone were to pass something besides the default bit generator (which I believe is a `PCG64`).

Reply:
I usually prefer having a proper check like `if not isisinstance(..., ...): ...` rather than using `try` `except`, but in this case I think your proposal it's way easier than implementing that check since the possible types of `seed` is a little bit complex.

But I'm not sure about storing the `self._seed` as the `seed_seq` of the generator. I think it could create the confusion of assuming that the random numbers generated in the class are done by setting a new rng with that seed and from the beginning of the random sequence.
While in reality, the random numbers are generated from the current random sequence in the generator that was passed. And in the end, we can still access that seed with `directive_obj.seed.bit_generator.seed_seq`.

I also think we should assign the passed value to `self._seed` even if it's a random number generator. This keeps the interface simple and less confusing than having something like:

```python
rng = np.random.default_rng(seed=42)
directive_obj = DirectiveClass(..., seed=rng)
print(directive_obj.seed)
```
```
SeedSequence(
    entropy=42,
)
```

What do you think?

Reply:
I'm more so looking for a way to restart (as it were) the process further down the line in a repeatable manner. So at the very least I think we should store the random number generator on the classes where appropriate. You can access the current state of the `BitGenerator` for a `Generator` as:
```python
generator.bit_generator.state
```

Which could then be used to recreate the `BitGenerator` at whatever state you were interested in.

Maybe it just doesn't make sense to store the "seed" property on classes at all when using generators. We would only allow a seed to be passed to the `__init__`.

Reply:
I think the capabilities for restarting the random state are already there. You can either reuse the same seed (as `int`), or redefine your Generator with the same seed and with the same state in the random sequence. For example:

```python
seed = 42
directive_obj = DirectiveClass(..., seed=seed)

# Run an inversion and use the directive
...

# Restart random state
directive_obj.seed = seed

# Reuse the directive after resetting the random state
...
```

And I think we should still store the `seed` as an attribute with the same value passed to the constructor, whether it is an `int` or a Generator, to simplify the interface as I pointed out in the last comment.

Nonetheless, I think your proposal would imply a change in the current behavior and/or interface of these classes, which I think is kind of out of the scope of this PR. I'd love to continue this discussion, but maybe we can find a compromise on this PR to move forward and then discuss how to improve the interface for these classses. Do you agree?

Reply:
That's a good point that the functionality is already there, although it's not directly from simpeg. I was mostly imagining the case where a user initializes the seed on a class instance with an integer, but then would want to access the created generator itself. I'm not saying we shouldn't store the seed, just was bringing it up as a potential option.

Reply:
There is a lot of power in the random number generators! For now, my 2 cents would be to keep it simple. I suspect that many users won't be aware of the workings of random number generators, so if you set a number and then ask for what you set and see something like 
```
SeedSequence(
    entropy=42,
)
```
that might be a bit of an intimidating animal to encounter 🐻 😄